libs/core/algorithms/tests/unit/algorithms/uninitialized_relocate_backward.cpp

```diff
@@ -0,0 +1,308 @@
+    int data;
+
+    explicit non_trivially_relocatable_struct(int data)
+      : data(data)
+    {
+        count++;
+    }
+    // mark as noexcept to enter simpler relocation path
+    non_trivially_relocatable_struct(
+        non_trivially_relocatable_struct&& other) noexcept
+      : data(other.data)
+    {
+        move_count++;
+        count++;
+    }
+    ~non_trivially_relocatable_struct()
+    {
+        dtor_count++;
+        count--;
+    }
+
+    // making sure the address is never directly accessed
+    friend void operator&(non_trivially_relocatable_struct) = delete;
+};
+int non_trivially_relocatable_struct::count = 0;
+int non_trivially_relocatable_struct::move_count = 0;
+int non_trivially_relocatable_struct::dtor_count = 0;
+
+static_assert(!is_trivially_relocatable_v<non_trivially_relocatable_struct>);
+
+struct non_trivially_relocatable_struct_throwing
+{
+    static int count;
+    static int move_count;
+    static int dtor_count;
+
+    int data;
+
+    explicit non_trivially_relocatable_struct_throwing(int data)
+      : data(data)
+    {
+        count++;
+    }
+    // do not mark as noexcept to enter try-catch relocation path
+    non_trivially_relocatable_struct_throwing(
+        non_trivially_relocatable_struct_throwing&& other)
+      : data(other.data)
+    {
+        if (move_count == K)
+        {
+            throw 42;
+        }
+        move_count++;
+
+        count++;
+    }
+    ~non_trivially_relocatable_struct_throwing()
+    {
+        dtor_count++;
+        count--;
+    }
+
+    // making sure the address is never directly accessed
+    friend void operator&(non_trivially_relocatable_struct_throwing) = delete;
+};
+
+int non_trivially_relocatable_struct_throwing::count = 0;
+int non_trivially_relocatable_struct_throwing::move_count = 0;
+int non_trivially_relocatable_struct_throwing::dtor_count = 0;
+
+static_assert(
+    !is_trivially_relocatable_v<non_trivially_relocatable_struct_throwing>);
+
+int hpx_main()
+{
+    {
+        void* mem1 = std::malloc(N * sizeof(trivially_relocatable_struct));
+        void* mem2 = std::malloc(N * sizeof(trivially_relocatable_struct));
+
+        HPX_TEST(mem1 && mem2);
+
+        trivially_relocatable_struct* ptr1 =
+            static_cast<trivially_relocatable_struct*>(mem1);
+        trivially_relocatable_struct* ptr2 =
+            static_cast<trivially_relocatable_struct*>(mem2);
+
+        HPX_TEST(trivially_relocatable_struct::count == 0);
+        HPX_TEST(trivially_relocatable_struct::move_count == 0);
+        HPX_TEST(trivially_relocatable_struct::dtor_count == 0);
+
+        for (int i = 0; i < N; i++)
+        {
+            hpx::construct_at(ptr1 + i, 1234);
+            hpx::construct_at(ptr2 + i, 5678);
+        }
+
+        // N objects constructed
+        HPX_TEST(trivially_relocatable_struct::count == 2 * N);
+
+        // relocate M objects to ptr2
+        uninitialized_relocate_backward(ptr1, ptr1 + M, ptr2 + N);
```

Comment:
Perhaps you should precede this line with `std::destroy(ptr2 + M - N, ptr2 + N)` so that you're not relocating-over-top-of any live objects.

Your `trivially_relocatable_struct` is trivially destructible, so this doesn't matter to the physical outcome. But, since you're so careful to (trivially) `std::destroy` certain subranges on lines 178–179 below, you might as well do the same here.